pkg/controller/scheduledscaling/scheduled_scaling.go

```diff
@@ -183,30 +191,220 @@ func (c *Controller) runOnce(ctx context.Context) error {
+				maxValue = activeSchedule.Value
+			}
+		}
+		currentActiveSchedules[schedule.Namespace+"/"+schedule.Name] = maxValue
+	}
+
+	for _, schedule := range clusterschedules {
+		activeSchedules, err := c.activeSchedules(schedule.Spec)
+		if err != nil {
+			log.Errorf("Failed to check for active schedules in ClusterScalingSchedule %s: %v", schedule.Name, err)
+			continue
+		}
+
+		if len(activeSchedules) == 0 {
+			continue
+		}
+
+		maxValue := int64(0)
+		for _, activeSchedule := range activeSchedules {
+			if activeSchedule.Value > maxValue {
+				maxValue = activeSchedule.Value
+			}
+		}
+		currentActiveSchedules[schedule.Name] = maxValue
+	}
+
+	return currentActiveSchedules
+}
+
+// adjustHPAScaling adjusts the scaling for a single HPA based on the active
+// scaling schedules. An adjustment is made if the current HPA scale is below
+// the desired and the change is within the HPA tolerance.
+func (c *Controller) adjustHPAScaling(ctx context.Context, hpa *autoscalingv2.HorizontalPodAutoscaler, activeSchedules map[string]int64) error {
+	current := int64(hpa.Status.CurrentReplicas)
+	if current == 0 {
+		return nil
+	}
+
+	var highestExpected int64
+	var highestObject autoscalingv2.CrossVersionObjectReference
+	for _, metric := range hpa.Spec.Metrics {
+		if metric.Type != autoscalingv2.ObjectMetricSourceType {
+			continue
+		}
+
+		switch metric.Object.DescribedObject.Kind {
+		case "ClusterScalingSchedule", "ScalingSchedule":
+		default:
+			continue
+		}
+
+		scheduleName := metric.Object.DescribedObject.Name
+
+		target := int64(metric.Object.Target.AverageValue.MilliValue() / 1000)
+		if target == 0 {
+			continue
+		}
+
+		var value int64
+		switch metric.Object.DescribedObject.Kind {
+		case "ScalingSchedule":
+			value = activeSchedules[hpa.Namespace+"/"+scheduleName]
+		case "ClusterScalingSchedule":
+			value = activeSchedules[scheduleName]
+		}
+
+		expected := int64(math.Ceil(float64(value) / float64(target)))
+		if expected > highestExpected {
+			highestExpected = expected
+			highestObject = metric.Object.DescribedObject
+		}
+	}
+
+	highestExpected = int64(math.Min(float64(highestExpected), float64(hpa.Spec.MaxReplicas)))
+
+	var change float64
+	if highestExpected > current {
+		change = (float64(highestExpected) - float64(current)) / float64(current)
+	}
+
+	if change > 0 && change <= c.hpaTolerance {
+		err := c.scaler.Scale(ctx, hpa, int32(highestExpected))
+		if err != nil {
+			reference := fmt.Sprintf("%s/%s/%s", hpa.Spec.ScaleTargetRef.Kind, hpa.Namespace, hpa.Spec.ScaleTargetRef.Name)
+			log.Errorf("Failed to scale target %s for HPA %s/%s: %v", reference, hpa.Namespace, hpa.Name, err)
+			return nil
+		}
+
+		scheduleRef := highestObject.Name
+		if highestObject.Kind == "ScalingSchedule" {
+			scheduleRef = hpa.Namespace + "/" + scheduleRef
+		}
+
+		c.recorder.Eventf(
+			hpa,
+			corev1.EventTypeNormal,
+			"ScalingAdjusted",
+			"Adjusted replicas %d -> %d based on schedule scaling metric: %s '%s'",
```

Comment:
Maybe it makes it more readable if we prepend the schedule name to the event message like:

```
Scaling schedule '%s' adjusted replicas %d -> %d based on metric: %s
```

Or something similar, wdyt?

Reply:
Changed